Feature/seunghojeong

seungho-jeong opened the pull request:
>wecode PR message template 및 체크 리스트 입니다.
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- URL 값에 적절한 필드사용 및 ManyToMany 필드 관련 세부조정사항 반영
		 
## 수정 사항들 자세한 내용
- URL 필드 속성 변경(CharField => URLField)
- Tag 클래스에 있는 ManyToManyField 다른 클래스로 변경(보다 주체가 되는 클래스쪽으로 이동)
- User 클래스의 자기참조 필드 관련 중간 테이블에 고유조합 속성 부여(Unique_together)

## 기타 질문 및 특이 사항
- 리뷰 주신 내용을 참고하여 Unique_together에 대해서 Goolging해보았습니다.
Unique_together가 ManyToMany 중간 테이블의 Field 조합의 고유성을 확보하기 위해 사용하는 것 같은데
이것이 어떤 경우를 대비하여 사용하는지 쉽게 이해되지 않습니다.
그렇다면 자기참조가 아닌 일반적인 M-to-M 관계의 경우에도 많은 데이터를 처리한 다는 가정하에 동일한 옵션을 주어야 하는지요.

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

seungho-jeong committed:
Add: 회원관리(user) App. 추가
- 회원정보 관리 DB용 Models.py 작성
- views.py 내 회원가입, 로그인(token 생성) 코드 작성
- urls.py 내 회원가입, 로그인 엔드포인트 생성
Modify: user App. 추가에 따른 settings.py 파일 수정
- INSTALLED_APP 내 "user" 추가

seungho-jeong committed:
Modify: Models.py 필드명 일부 수정, Git Push 제외대상 추가
- (Models.py)필드명 변경, short_introduce => introduction
- (Models.py)필드명 변경, profile_image => profile_image_url
- (Models.py)필드 신규추가, created_at
- (Models.py)필드 신규추가, updated_at
- (.gitignore)my_uploader.py 추가

soheon-lee commented:
@Dan-Jeong 승호님 현재 models.py 만 merge가능할 것 같아서 views.py에 있는 코드는 잠깐 다른 곳에 백업해두고, 파일 내용 지우고 올려주세요~

seungho-jeong committed:
Remove: Views.py 삭제

seungho-jeong committed:
Add: work(아트워크) App. 생성 및 모델 작성
- 아트워크(work) App. 신규생성
- 작품, 배경화면, 댓글, 태그 관련 클래스 작성
- ERD: https://aquerytool.com:443/aquerymain/index/?rurl=581ffee9-9665-4a2d-bd14-5caeeed378a8&
- P/W: 61x206
Modity: User 클래스 및 Settings.py 수정
- User ~ Follow 자기참조를 위하여 Follow 클래스 생성
- Settings.py 내 work App 추가

lordmyshepherd commented on work/models.py:
```diff
@@ -0,0 +1,110 @@
+
+class WorkImage(models.Model):
+    work            = models.ForeignKey(Work, on_delete = models.CASCADE)
+    image_url       = models.CharField(max_length = 1000)
```
URLField를 사용해보시면 어떨까요?

seungho-jeong replied:
리뷰 코멘트 반영하여 코드 수정하였습니다.

lordmyshepherd commented on work/models.py:
```diff
@@ -0,0 +1,110 @@
+
+class Tag(models.Model):
+    name            = models.CharField(max_length = 40)
+    category        = models.ManyToManyField(Category, through = "CategoryToTag", related_name = "categories")
```
many to many field는 tag 보다는 category 쪽에 위치하는게 좋습니다.
이유는 category에서 연관된 tag를 조회하는 경우가, tag에서 category를 조회하는 경우보다 많을 것으로 예상되기 때문입니다~

seungho-jeong replied:
보다 주체가 되는 쪽에 ManyToMany 필드를 넣어주었습니다.

lordmyshepherd commented on work/models.py:
```diff
@@ -0,0 +1,110 @@
+class Tag(models.Model):
+    name            = models.CharField(max_length = 40)
+    category        = models.ManyToManyField(Category, through = "CategoryToTag", related_name = "categories")
+    work            = models.ManyToManyField(Work, through = "WorkToTag", related_name = "works")
```
마찬가지 입니다~

seungho-jeong replied:
보다 주체가 되는 쪽에 ManyToMany 필드를 넣어주었습니다.

lordmyshepherd commented on user/models.py:
```diff
@@ -0,0 +1,24 @@
+    profile_image_url   = models.CharField(max_length = 1000, null = True)
+    created_at          = models.DateTimeField(auto_now_add = True)
+    updated_at          = models.DateTimeField(auto_now = True)
+    following           = models.ManyToManyField(
```
아래 코드를 참고해서 수정해 보시겠어요?
```
class User(models.Model):
    korean_name       = models.CharField(max_length=45) 
    password          = models.CharField(max_length=200)
    friends           = models.ManyToManyField('self', through='Friend', related_name='followers', symmetrical=False)

    class Meta:
        db_table = 'users'

class Friend(models.Model):
    from_user = models.ForeignKey('User', on_delete=models.CASCADE, related_name='to_user')
    to_user   = models.ForeignKey('User', on_delete=models.CASCADE, related_name='from_user')

   class Meta:
        unique_together = ('from_user', 'to_user')
        db_table        = 'friends'
```

seungho-jeong replied:
리뷰 코멘트 반영하여 코드 수정하였으나 궁금한 점이 있어 코멘트 남깁니다.

Question.
리뷰 주신 내용을 참고하여 Unique_together에 대해서 Goolging해보았습니다.
Unique_together가 ManyToMany 중간 테이블의 Field 조합의 고유성을 확보하기 위해 사용하는 것 같은데
이것이 어떤 경우를 대비하여 사용하는지 쉽게 이해되지 않습니다.
그렇다면 자기참조가 아닌 일반적인 M-to-M 관계의 경우에도 많은 데이터를 처리한 다는 가정하에 동일한 옵션을 주어야 하는지요.

soheon-lee replied:
일반적인 M-to-M 필드에서도 그 조합이 고유해야하는 경우에는 해당 옵션을 줄 수 있습니다 :)

seungho-jeong committed:
Modify: 리뷰 코멘트 사항 반영
- URL 필드 속성 변경(CharField => URLField)
- Tag 클래스에 있는 ManyToManyField 다른 클래스로 변경(보다 주체가 되는 클래스쪽으로 이동)
- User 클래스의 자기참조 필드 관련 중간 테이블에 고유조합 속성
  부여(Unique_together)

soheon-lee commented on work/models.py:
```diff
@@ -0,0 +1,113 @@
+
+class Tag(models.Model):
+    name            = models.CharField(max_length = 40)
+    
```
불필요한 간격은 삭제해주세요

seungho-jeong replied:
클래스별 필드명 띄어쓰기 적절한 간격으로 수정하였습니다.

soheon-lee commented:
@Dan-Jeong 아 그리고 브랜치 이름은 기능 단위로 작성합니다 :) 이후 부터는 이름 말고 기능 단위로 올려주세요~

seungho-jeong committed:
Modify: 리뷰 코멘트(불필요한 Indention 삭제) 반영

lordmyshepherd commented:
LGTM 👍
